code/Control/UserDefinedFormController.php

```diff
@@ -511,6 +533,250 @@ public function finished()
+
+
+        $options = new OptionsetField('FolderOptions', 'Folder options', [
+            "new" => 'Create a new folder (recommended)',
+            "existing" => 'Use an existing folder'
+        ], "new");
+        $fields->push($options);
+
+
+        $treeView = TreeDropdownField::create(
+            'FolderID',
+            '',
+            Folder::class
+        )->setValue((int) $folderId);
+        $treeView->addExtraClass('pt-1');
+        $treeView->setDescription(static::getFolderPermissionString($folder));
+        $fields->push($treeView);
+
+
+        $fields->push(new HiddenField('ID', 'ID', $editableFormField->ID));
+
+
+        $submitAction = new FormAction('confirmfolder', "Confirm");
+        $submitAction->setUseButtonTag(false);
+        $submitAction->addExtraClass('btn');
+        $submitAction->addExtraClass('btn-primary');
+
+        $cancelAction = new FormAction('cancel', "Cancel");
+        $cancelAction->setUseButtonTag(false);
+        $cancelAction->addExtraClass('btn');
+
+
+        $form = new Form($this, 'Form', $fields, new FieldList($submitAction, $cancelAction));
+        $form->setFormAction("UserDefinedFormController/confirmfolder");
+
+
+        if (!$editableFormField instanceof EditableFileField) {
+            $editableFormField->setClassName(EditableFileField::class);
+            $editableFormField->write();
+        }
+        $treeView->setSchemaData([
+            'data' => [
+                'urlTree' => "admin/pages/edit/EditForm/$userFormID/field/Fields/item/$id/ItemEditForm/field/FolderID/tree"
+            ]
+        ]);
+
+
+        // create the schema response
+        $parts = $this->getRequest()->getHeader(LeftAndMain::SCHEMA_HEADER);
+        $schemaID = $this->getRequest()->getURL();
+        $data = FormSchema::singleton()
+            ->getMultipartSchema($parts, $schemaID, $form);
+
+        // return the schema response
+        $response = new HTTPResponse(json_encode($data));
+        $response->addHeader('Content-Type', 'application/json');
+        return $response;
+    }
+
+    /**
+     * Sets the selected folder as the upload folder for an EditableFileField
+     * @return HTTPResponse
+     * @throws ValidationException
+     */
+    public function confirmfolder()
+    {
+        $request = $this->getRequest();
+
+        // retrieve the EditableFileField
+        $id = $request->requestVar('ID');
+        /** @var EditableFileField $editableFileField */
+        $editableFormField = EditableFormField::get()->byID($id);
+        if (!$editableFormField) {
+            $editableFormField = Versioned::get_by_stage(EditableFormField::class, Versioned::DRAFT)->byID($id);
+        }
+        // change the class if it is incorrect
+        if (!$editableFormField instanceof EditableFileField) {
+            $editableFormField->setClassName(EditableFileField::class);
+        }
+        $editableFileField = $editableFormField;
+
+        if (!$editableFileField->canView()) {
```

Comment:
Shouldn't the check be `canEdit` here? We are writing the field a bit further down.